napari/layers/shapes/_shape_list.py

```diff
@@ -353,6 +352,170 @@ def add(
             # Set z_order
             self._update_z_order()
 
+    def add_multiple(
+        self,
+        shapes,
+        face_colors=None,
+        edge_colors=None,
+        z_refresh=True,
+    ):
+        """Add multiple shapes at once (faster than adding them one by one)
+
+        Parameters
+        ----------
+        shapes : iterable of Shape
+            Each Shape must be a subclass of Shape, one of "{'Line', 'Rectangle',
+            'Ellipse', 'Path', 'Polygon'}"
+        face_colors : iterable of face_color
+        edge_colors : iterable of edge_color
+        z_refresh : bool
+            If set to true, the mesh elements are reindexed with the new z order.
+            When shape_index is provided, z_refresh will be overwritten to false,
+            as the z indices will not change.
+            When adding a batch of shapes, set to false  and then call
+            ShapesList._update_z_order() once at the end.
+
+        TODO: Currently shares a lot of code with `add()`, with the
+        difference being that `add()` supports inserting shapes at a specific
+        `shape_index`, whereas `add_multiple` will append them as a full batch
+        """
+
+        def _make_index(length, shape_index, cval=0):
+            """Same but faster than `np.repeat([[shape_index, cval]], length, axis=0)`"""
+            index = np.empty((length, 2), np.int32)
+            index.fill(cval)
+            index[:, 0] = shape_index
+            return index
+
+        all_z_index = list()
+        all_vertices = list()
+        all_index = list()
+        all_mesh_vertices = list()
+        all_mesh_vertices_centers = list()
+        all_mesh_vertices_offsets = list()
+        all_mesh_vertices_index = list()
+        all_mesh_triangles = list()
+        all_mesh_triangles_index = list()
+        all_mesh_triangles_colors = list()
+
+        m_mesh_vertices_count = len(self._mesh.vertices)
+
+        if face_colors is None:
+            face_colors = np.tile(np.array([1, 1, 1, 1]), (len(shapes), 1))
+        else:
+            face_colors = np.asarray(face_colors)
+
+        if edge_colors is None:
+            edge_colors = np.tile(np.array([0, 0, 0, 1]), (len(shapes), 1))
+        else:
+            edge_colors = np.asarray(edge_colors)
+
+        if not len(face_colors) == len(edge_colors) == len(shapes):
+            raise ValueError(
+                trans._(
+                    'shapes, face_colors, and edge_colors must be the same length',
+                    deferred=True,
+                )
+            )
+
+        if not all(issubclass(type(shape), Shape) for shape in shapes):
+            raise ValueError(
+                trans._(
+                    'all shapes must be subclass of Shape',
+                    deferred=True,
+                )
+            )
+
+        for shape, face_color, edge_color in zip(
+            shapes, face_colors, edge_colors
+        ):
```

Comment:
I have to agree with @kevinyamauchi, this function is so long. I really have trouble following it. It would be great if it was refactored into some helper functions.

Reply:
Yeah, I agree. It's still like this as I only "translated" the existing `add()` function into a version accepting multiple shapes, without attempting to refactor the details in fear of possible side effects